jetbrains-core/src/software/aws/toolkits/jetbrains/core/executables/ExecutableManager2.kt

```diff
@@ -0,0 +1,88 @@
+import com.intellij.openapi.project.Project
+import com.intellij.util.ExceptionUtil
+import software.aws.toolkits.jetbrains.utils.assertIsNonDispatchThread
+import software.aws.toolkits.jetbrains.utils.runUnderProgressIfNeeded
+import software.aws.toolkits.resources.message
+import java.nio.file.Path
+import java.nio.file.Paths
+
+class ExecutableManager2 {
+    private val versionCache = ExecutableVersionCache()
+
+    fun <T : ExecutableType2<V>, V : Version> getExecutable(type: T): Executable<T>? = getExecutablePath(type)?.let { getExecutable(type, it) }
+    fun <T : ExecutableType2<V>, V : Version> getExecutable(type: T, path: Path): Executable<T> = Executable(type, path)
+
+    private fun <T : ExecutableType2<*>> getExecutablePath(type: T): Path? {
+        // Check if user gave a custom path
+        ExecutableSettings.getInstance().getExecutablePath(type)?.let {
+            return Paths.get(it)
+        }
+
+        return detectExecutable(type)
+    }
+
+    private fun <T> detectExecutable(type: T): Path? = if (type is AutoResolvable) {
+        type.resolve()
+    } else {
+        null
+    }
+
+    fun <T : Version> determineVersion(executable: Executable<ExecutableType2<T>>): T {
+        assertIsNonDispatchThread()
+
+        return when (val result = versionCache.getValue(executable)) {
+            is Result.Failure -> {
+                throw RuntimeException(message("failed to identify version of {0}", executable.type.displayName), result.reason)
+            }
+            is Result.Success<T> -> {
+                result.version
+            }
+        }
+    }
+
+    fun <T : Version> validateCompatability(
+        project: Project?,
+        executable: Executable<ExecutableType2<T>>?,
+        additionalStrictRange: List<VersionRange<T>>? = null
+    ): Compatability {
+        if (executable == null) {
+            return Compatability.NotInstalled()
+        }
+
+//        return runUnderProgressIfNeeded(project, message("version.progress.title"), false) {
+        return runUnderProgressIfNeeded(project, "Validating CLI", false) {
+            determineCompatability(executable, additionalStrictRange)
+        }
+    }
+
+    private fun <T : Version> determineCompatability(executable: Executable<ExecutableType2<T>>, additionalStrictRange: List<VersionRange<T>>?): Compatability {
+        assertIsNonDispatchThread()
+
+        val version = when (val cacheResult = versionCache.getValue(executable)) {
+            is Result.Failure -> return Compatability.NotInstalled(ExceptionUtil.getMessage(cacheResult.reason))
+            is Result.Success -> cacheResult.version
+        }
+
+        val baseVersionCompatability = isVersionValid(version, executable.type.supportedVersions())
+        if (baseVersionCompatability != Compatability.Valid) {
+            return baseVersionCompatability
+        }
+
+        additionalStrictRange?.let {
+            return isVersionValid(version, additionalStrictRange)
+        }
```

Comment:
Our use case for this is "this one feature needs 1.X+2, but rest of toolkit needs 1.X", so maybe this should actually be `additionalMinVersion: Version`?

Reply:
The reason I was hesitant was then it needs to worry about the max version